import psycopg2
from typing import List
from errors import CacheException, OtherProvidersException
import json
import requests



class FilesCacheManager:
    def __init__(self, db_params:json):
        self.db_params = db_params

    
    def get_file_details_from_db(self, file_hash:str):
        try:
            with psycopg2.connect(**self.db_params) as conn:
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM android_apps_cache_files WHERE file_hash = %s", (file_hash,))
                existing_item = cursor.fetchone()
                cursor.close()
                return existing_item
        except Exception as e:
            raise CacheException(f"Error occurred while run get_file_details_from_db: {e}")



    def insert_results_to_db(self, file_hash:str, verdict:str, probability:float, permissions:str):
        try:    
            with psycopg2.connect(**self.db_params) as conn:
                cursor = conn.cursor()
                query = f"INSERT INTO android_apps_cache_files (file_hash, verdict, probability, permissions) VALUES ('{file_hash}','{verdict}',{probability},'{permissions}');"
    
                cursor.execute(query)
                conn.commit()
                cursor.close()
        except Exception as e:
            raise CacheException(f"Error occurred while run insert_results_to_db: {e}")


class OtherProviders:
   
    def scan_with_virustotal(url:str, params:json):
        try:
            response = requests.get(url, params=params)
            if response.status_code == 200:
                report = response.json()
                if report['positives'] > 0:
                    return 'Malicious'
                else:
                    return 'Benign'
            else:
                raise requests.RequestException(f"status code: {response.status_code}")
        
        except Exception as e:
            raise OtherProvidersException(f"Error occurred while try to scan the file with Virus Total: {e}")